### PR TITLE
bpo-39308: Make _tp_cache typed

### PR DESCRIPTION
**Pull request by kornicameister**
https://bugs.python.org/issue39308

Automerge-Triggered-By: @gvanrossum

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@kornicameister

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by kornicameister** [comment]
ok, haven't exactly read entire contributing docs and missed `News` bit, I will deal with that shortly.
I am more worried about lack of CLA which I have signed just yesterday :/

**Review by gvanrossum — APPROVED** [review]
Looks good. CLA usually takes a business day.

**Comment by kornicameister** [comment]
@gvanrossum alright, so let's wait for CLA to be processed. I assume my job here is done and bot will simply change label if CLA gets in.

cheers :+1:

**Comment by ilevkivskyi** [comment]
I am a bit worried about performance impact. This is relevant only for `Literal[...]` but will affect _all_ generics like `List[int]`. If the performance impact is small, then it is OK, otherwise I would specifically used the typed cache version only for `Literal[...]`.

**Comment by ilevkivskyi** [comment]
(Removed the auto-merge label until perf question is resolved).

**Comment by kornicameister** [comment]
@ilevkivskyi seems a bit contradicting ;), performance issue and caching. But that's worth checking anyway, I will do such comparison later on.

**Comment by ilevkivskyi** [comment]
What is contradicting here? Typed cache lookup will be slower (because we need to compare the types), the question is how much slower.

**Comment by gvanrossum** [comment]
Looking at the [source](https://github.com/python/cpython/blob/master/Lib/functools.py#L461), the difference in the cache is that with `typed=False`, the cache key used for `Literal[1]` will be the integer object `1`, while with `typed=True` it will construct a tuple `(1, int)`. Or, more likely, since `@_tp_cache` is applied to a *method* (`__getitem__`), the cache key without will be the tuple `(self, 1)` and the cache key with will be `(self, 1, Literal, int)`. So cache key construction will be a bit slower and cache lookup will too. Using timeit it might be easy for the OP to check whether this is significant. (Note that the "real" code is written in C, and I didn't look at it, but IIRC it faithfully follows the Python version linked above.)

**Comment by kornicameister** [comment]
Results:

|stmt|typed=False|typed=True|2 caches|
|-----|-----------------|---------------|---------|
|`t.Literal[1]`|0.46885ms|0.53354ms|0.52857ms|
|`t.Literal[1]`|0.43094ms|0.49489ms|0.50706ms|
|`(t.Literal[1], t.Literal[1], t.Literal[0], t.Literal[0])`|1.71090ms|1.99839ms|2.08259ms|
|`t.Literal[True]`|0.42652ms|0.46590ms|0.51885ms|
|`t.Literal[True, False]`|0.42159ms|0.46590ms|0.53356ms|
|`t.Literal[0]`|0.41263ms|0.46680ms|0.51893ms|
|`t.Literal[0, 1]`|0.42797ms|0.47858ms|0.54927ms|
|`t.List[t.Literal[1,2,3]]`|1.68909ms|1.85799ms|1.88905ms|
|`t.Union[float, complex]`|0.57196ms|0.55289ms|0.58154ms|
|`t.Union[str, bytes, float, complex, t.Literal[1]]`|1.57654ms|1.68061ms|1.74077ms|
|`t.NewType("NT", int)`|0.31505ms|0.30028ms|0.29731ms|
|`t.Literal[1,2,3,4,5,6,7,8]`|0.44533ms|0.49266ms|0.54586ms|
|`t.Dict[t.AnyStr, complex]`|0.71057ms|0.80595ms|0.78011ms|
|`t.Set[t.AnyStr]`|0.67253ms|0.72656ms|0.68497ms|
|`t.Set[t.Union[float, int]]`|1.82547ms|1.99398ms|2.18703ms|
|`t.AbstractSet[t.Union[float, int]]`|1.84454ms|1.96812ms|2.12587ms|
|`t.Sequence[t.Union[float, int]]`|1.88198ms|1.98809ms|2.01617ms|
|`t.NamedTuple("A", x=int)`|56.12496ms|55.05577ms|56.93304ms|
|`t.NamedTuple("B", x=int,y=float)`|60.79059ms|62.23118ms|63.67331ms|
|`t.NamedTuple("C", x=int,y=float,z=complex)`|65.46186ms|63.83327ms|66.34171ms|


Scripts:

```python
import typing as t
import timeit as tt
import time
import operator
import functools

stmts = (
    't.Literal[1]',
    't.Literal[1]',
    '(t.Literal[1], t.Literal[1], t.Literal[0], t.Literal[0])',
    't.Literal[True]',
    't.Literal[True, False]',
    '(t.Literal[True, False], t.Literal[True, False], t.Literal[True, False], t.Literal[True, False], t.Literal[True, False])',
    't.Literal[0]',
    't.Literal[0, 1]',
    't.List[t.Literal[1,2,3]]',
    't.Union[float, complex]',
    't.Union[str, bytes, float, complex, t.Literal[1]]',
    't.NewType("NT", int)',
    't.Literal[1,2,3,4,5,6,7,8]',
    't.Dict[t.AnyStr, complex]',
    't.Set[t.AnyStr]',
    't.Set[t.Union[float, int]]',
    't.AbstractSet[t.Union[float, int]]',
    't.Sequence[t.Union[float, int]]',
    't.NamedTuple("A", x=int)',
    't.NamedTuple("B", x=int,y=float)',
    't.NamedTuple("C", x=int,y=float,z=complex)',
)
repeats = 300.0
exec_number = 1000

for stmt in stmts:
    times=tt.repeat(
        stmt,
        globals={
            't': t,
        },
        number=exec_number,
        repeat=int(repeats),
        timer=time.perf_counter,
    )
    avg = functools.reduce(operator.add, times) / repeats
    print(f'{stmt} => {avg*1000.0:.5f}ms')
```

indeed, having `typed=False` results in performance impact but worst results can be observed when there are two caches one with `typed=True` and the other with `typed=False` i.e.:

```python
normal_cache = functools.lru_cache(typed=False)(func)
literal_cached = functools.lru_cache(typed=True)(func)

_cleanups.append(normal_cache.cache_clear)
_cleanups.append(literal_cached.cache_clear)
```

**Comment by gvanrossum** [comment]
Can you clarify what you mean by two caches?

> --
--Guido (mobile)

**Comment by kornicameister** [comment]
> a bit worried about performance impact. This is relevant only for `Literal[...]` but will affect _all_ generics like `List[int]`. If the performance impact is small, then it is OK, otherwise I would specifically used the typed cache version only for `Literal[...]`.

@ilevkivskyi mentioned about having cache `typed=True` cache just for `Literal`.

**Comment by kornicameister** [comment]
I was hoping that maybe there will be a possibility to affect `__eq__` or `__hash__` for `Literal` but given how  it is constructed, I don't see how it can be achieved at reasonable cost.

**Comment by ilevkivskyi** [comment]
TBH, this doesn't make much sense. If you only change the behavior of the cache for `Literal[...]`, then how this can affect the performance of existing unchanged code for `List[int]`? You likely just see the noise (btw there is no variance in your table).

**Comment by kornicameister** [comment]
perhaps the way I've done that affected results, I've simply put a ternary operator inside of function that is wrapped in `_tp_cache` to have something like: `return literal_cached(*args,**kwargs) if args[0] is Literal else cached(*args, **kwargs**)`.

@ilevkivskyi was that what you had in mind?

**Comment by ilevkivskyi** [comment]
No, you should remove the `Literal` branch at the end of `_SpecialForm.__getitem__()`, override `__getitem__()` (used the removed branch and typed cache) in a subclass (lets say `_TypedSpecialForm` or `_LiteralSpecialForm`), and make `Literal` an instance of the latter.

**Comment by kornicameister** [comment]
@ilevkivskyi thx for that tip, that will save us some time during reviewing :+1: 
Before a patch is posted, a little heads up. I've been trying to utilize existing `_tp_cache` function and just make it aware of `typed: bool` argument that defaults to `False`. Following error is what stopped me from doing just that:

```function() argument 1 must be code, not str```.

I couldn't track where does it come from so decided to just make two decorators one for `typed=True` and the other for `typed=False`.

File: Lib/typing.py [hunk 1]
```diff
@@ -363,11 +380,18 @@ def __getitem__(self, parameters):
...
+class _TypedSpecialForm(_SpecialForm, _root=True):
+
+    @_typed_tp_cache
+    def __getitem__(self, parameters):
         if self._name == 'Literal':
             # There is no '_type_check' call because arguments to Literal[...] are
             # values, not types.
             return _GenericAlias(self, parameters)
-        raise TypeError(f"{self} is not subscriptable")
+        return super().__getitem__(parameters)
```
Comment (kornicameister) [inline Lib/typing.py:394]: Not quite sure if that's acceptable though.

**Comment by gvanrossum** [comment]
I just realized that adding `typed=True` doesn't fix the issue completely, since e.g. `Literal[1, 1]` and `Literal[1, True]` still have the same cache key. Example:
```
>>> from typing import Literal
>>> Literal[1]
typing.Literal[1]
>>> Literal[True]
typing.Literal[True]
>>> Literal[1]
typing.Literal[1]
>>> Literal[1, 1]
typing.Literal[1, 1]
>>> Literal[True, True]
typing.Literal[1, 1]
>>> 
```
This shows that for simple `==` clashes the bug is fixed (`Literal[1]` and `Literal[True]` are not confused) but for tuples it is still not fixed, because the type appended to the key by `functools._make_key()` is `tuple`.

I don't know how to fix this. Perhaps it is fine not to fix it at all? Analysis by mypy or other static type checkers is unaffected, since they don't use the runtime implementation of `typing.py` at all.

**Comment by kornicameister** [comment]
@gvanrossum not fixing it sounds like an option, true. I just wonder...I mean I stumbled into that just because I've tried to use runtime implementation.

TBH I was not even aware, up until now, that there are, sort of, two versions of `typing.py`. That's not an argument to struggle for a fix, it's just an argument that people might try to leverage runtime `typing.py` to implement certain type-based checks. 

Let me put it this way, just because I'm unsure how things are discussed. How about, you or @ilevkivskyi , put it somewhere on board so that people can discuss it. TBH I've already tried and nobody was able to give any ideas :(. In the meantime and since this affects Python `>=3.8`, I will try to come up with `lru_cache` that uses my own caching function. Worst case scenario, we might replace `_TypedSpecialForm` to `_LiteralSpecialForm` and how it done like so. 

3rd idea...how about there's no caching for `Literal`? `Literal` does not utilize any expensive type checks and so on, it is merely a value container, right? If that's the case, at least theoretically, maybe dropping a cache is an option as well?

**Comment by gvanrossum** [comment]
IMO that people misunderstand how static type checks work is not a reason to change things.

I don't think there's a way to fix this with a cache unless you're willing to write your own equality test (one that considers `1 != True` and recurses into tuples).

Ditching the cache for `Literal[]` doesn't sound like a great option, it would be a problem if people start to adopt using `Literal[]` in code that doesn't use `from __future__ import annotations`. Maybe in a few years.

**Comment by kornicameister** [comment]
Well, I am not going to argue with the reasoning. Not to mention that recursion does not exactly shouts performance. And it does not make sense to change it for selected few if cost exceeds any gains.

So yeah, we done? we're closing it down as `no fix` :D. I happy either way, I got a chance to learn few bits I was not aware of.

**Pull request closed without merging**